File: riveredge-backend/src/apps/kuaizhizao/schemas/receipt_notice.py
"""
收货通知单数据验证schema

采购通知仓库收货，不直接动库存。

Author: RiverEdge Team
Date: 2026-02-22
"""

from datetime import datetime, date
from typing import Optional, List
from pydantic import Field
from core.schemas.base import BaseSchema


# === 收货通知单 ===

class ReceiptNoticeBase(BaseSchema):
    """收货通知单基础schema"""
    notice_code: Optional[str] = Field(None, max_length=50, description="通知单编码（可选，不提供则自动生成）")
    purchase_order_id: int = Field(..., description="采购订单ID")
    purchase_order_code: str = Field(..., max_length=50, description="采购订单编码")
    supplier_id: int = Field(..., description="供应商ID")
    supplier_name: str = Field(..., max_length=200, description="供应商名称")
    supplier_contact: Optional[str] = Field(None, max_length=100, description="供应商联系人")
    supplier_phone: Optional[str] = Field(None, max_length=50, description="供应商电话")
    warehouse_id: Optional[int] = Field(None, description="入库仓库ID")
    warehouse_name: Optional[str] = Field(None, max_length=100, description="入库仓库名称")
    planned_receipt_date: Optional[date] = Field(None, description="计划收货日期")
    status: str = Field("待收货", max_length=20, description="通知状态")
    notes: Optional[str] = Field(None, description="备注")


class ReceiptNoticeCreate(ReceiptNoticeBase):
    """收货通知单创建schema"""
    items: Optional[List["ReceiptNoticeItemCreate"]] = Field(None, description="通知明细列表")


class ReceiptNoticeUpdate(ReceiptNoticeBase):
    """收货通知单更新schema"""
    notice_code: Optional[str] = Field(None, max_length=50, description="通知单编码")


class ReceiptNoticeResponse(ReceiptNoticeBase):
    """收货通知单响应schema"""
    id: int = Field(..., description="通知单ID")
    tenant_id: int = Field(..., description="租户ID")
    notified_at: Optional[datetime] = Field(None, description="通知仓库时间")
    purchase_receipt_id: Optional[int] = Field(None, description="采购入库单ID")
    purchase_receipt_code: Optional[str] = Field(None, max_length=50, description="采购入库单编码")
    total_quantity: float = Field(0, description="总数量")
    total_amount: float = Field(0, description="总金额")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


class ReceiptNoticeListResponse(ReceiptNoticeResponse):
    """收货通知单列表响应schema"""
    pass


# === 收货通知单明细 ===

class ReceiptNoticeItemBase(BaseSchema):
    """收货通知单明细基础schema"""
    material_id: int = Field(..., description="物料ID")
    material_code: str = Field(..., max_length=50, description="物料编码")
    material_name: str = Field(..., max_length=200, description="物料名称")
    material_spec: Optional[str] = Field(None, max_length=200, description="物料规格")
    material_unit: str = Field(..., max_length=20, description="物料单位")
    notice_quantity: float = Field(..., gt=0, description="通知数量")
    unit_price: float = Field(0, ge=0, description="单价")
    total_amount: Optional[float] = Field(None, ge=0, description="金额")
    purchase_order_item_id: Optional[int] = Field(None, description="采购订单明细ID")
    notes: Optional[str] = Field(None, description="备注")


class ReceiptNoticeItemCreate(ReceiptNoticeItemBase):
    """收货通知单明细创建schema"""
    pass


class ReceiptNoticeItemUpdate(ReceiptNoticeItemBase):
    """收货通知单明细更新schema"""
    pass


class ReceiptNoticeItemResponse(ReceiptNoticeItemBase):
    """收货通知单明细响应schema"""
    id: int = Field(..., description="明细ID")
    tenant_id: int = Field(..., description="租户ID")
    notice_id: int = Field(..., description="通知单ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


class ReceiptNoticeWithItemsResponse(ReceiptNoticeResponse):
    """收货通知单详情响应（含明细）"""
    items: List[ReceiptNoticeItemResponse] = Field(default_factory=list, description="通知明细列表")

File: riveredge-frontend/src/apps/kuaizhizao/services/receipt-notice.ts
/**
 * 收货通知单 API
 */

import { apiRequest } from '../../../services/api';

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface ReceiptNoticeCapabilities {
  update?: ActionCapability;
  delete?: ActionCapability;
  notify?: ActionCapability;
  withdraw?: ActionCapability;
  print?: ActionCapability;
}

export interface ReceiptNotice {
  id?: number;
  notice_code?: string;
  purchase_order_id?: number;
  purchase_order_code?: string;
  supplier_id?: number;
  supplier_name?: string;
  supplier_contact?: string;
  supplier_phone?: string;
  warehouse_id?: number;
  warehouse_name?: string;
  planned_receipt_date?: string;
  status?: string;
  notified_at?: string;
  purchase_receipt_id?: number;
  purchase_receipt_code?: string;
  total_quantity?: number;
  total_amount?: number;
  notes?: string;
  attachments?: Array<{ uid?: string; name?: string; url?: string }>;
  created_at?: string;
  updated_at?: string;
  lifecycle?: Record<string, unknown>;
  capabilities?: ReceiptNoticeCapabilities;
}

export interface ReceiptNoticeNotifyPreviewResponse {
  target_type: string;
  summary: string;
  notice_code?: string;
  warehouse_required?: boolean;
  warehouse_id?: number | null;
  items: Array<{
    item_id?: number;
    purchase_order_item_id?: number | null;
    material_code: string;
    material_name: string;
    quantity: number;
    pushed_quantity?: number;
    max_push_quantity?: number;
    notice_quantity?: number;
  }>;
  tip?: string;
  has_blocking_issues?: boolean;
  blocking_reason?: string | null;
  line_blocking_issues?: string[];
}

const RECEIPT_NOTICE_LIMIT_MAX = 1000;

export interface ReceiptNoticeListParams {
  skip?: number;
  limit?: number;
  status?: string;
  purchase_order_id?: number;
  supplier_id?: number;
  warehouse_id?: number;
  keyword?: string;
  notice_code?: string;
  purchase_order_code?: string;
  planned_start_date?: string;
  planned_end_date?: string;
  created_start_date?: string;
  created_end_date?: string;
  order_by?: string;
  /** 明细表格视图：附带通知明细 */
  include_items?: boolean;
}

export interface ReceiptNoticeListResult {
  data: ReceiptNotice[];
  total: number;
  success: boolean;
}

function clampReceiptNoticeLimit(limit: unknown): number | undefined {
  if (typeof limit !== 'number' || Number.isNaN(limit)) return undefined;
  return Math.max(1, Math.min(RECEIPT_NOTICE_LIMIT_MAX, Math.trunc(limit)));
}

export const receiptNoticeApi = {
  list: async (params?: ReceiptNoticeListParams): Promise<ReceiptNoticeListResult> => {
    const limit = clampReceiptNoticeLimit(params?.limit);
    const safeParams = limit != null ? { ...params, limit } : params;
    return apiRequest('/apps/kuaizhizao/receipt-notices', { method: 'GET', params: safeParams });
  },
  create: async (data: any) => apiRequest('/apps/kuaizhizao/receipt-notices', { method: 'POST', data }),
  update: async (id: string, data: any) =>
    apiRequest(`/apps/kuaizhizao/receipt-notices/${id}`, { method: 'PUT', data }),
  delete: async (id: string) => apiRequest(`/apps/kuaizhizao/receipt-notices/${id}`, { method: 'DELETE' }),
  get: async (id: string) => apiRequest(`/apps/kuaizhizao/receipt-notices/${id}`, { method: 'GET' }),
  previewNotify: async (id: string) =>
    apiRequest<ReceiptNoticeNotifyPreviewResponse>(`/apps/kuaizhizao/receipt-notices/${id}/notify/preview`, {
      method: 'GET',
    }),
  notify: async (id: string) =>
    apiRequest(`/apps/kuaizhizao/receipt-notices/${id}/notify`, { method: 'POST' }),
  statistics: async () =>
    apiRequest<{ total: number; pending: number; notified: number; received: number }>(
      '/apps/kuaizhizao/receipt-notices/statistics',
      { method: 'GET' },
    ),
  withdraw: async (id: string) =>
    apiRequest(`/apps/kuaizhizao/receipt-notices/${id}/withdraw`, { method: 'POST' }),
};
